### ISSUE TITLE
Add ability to configure auth in embedded lightblue for integration testing

### ISSUE THREAD

**Issue by alechenninger** [lightblue-platform/lightblue-rest#208]
I have some business logic I want to test which involves specific configurable users, which may require lightblue authentication to be enabled.

That being said, I'm currently experimenting if perhaps sending basic auth with the client "just works" anyway, and in that case that would be fine.

**Comment by alechenninger** [comment]
So messed around with a few things. The main problem I think is that with the embedded http server, this line...

```java
        this.httpServletRequest = ResteasyProviderFactory.getContextData(HttpServletRequest.class);
```

...always returns `null` which means we can never get the principal with the request.

I tried adding basic auth to lightblue client and added a security context to the embedded http server, but still no dice.

**Comment by alechenninger** [comment]
If you want to experiment, I added basic auth to the client: https://github.com/lightblue-platform/lightblue-client/pull/268

**Comment by paterczm** [comment]
For auth & authz, lightblue relies on the container. Integration tests use a simple server container (com.sun.net.httpserver.HttpServer). It [can be setup to require basic auth](http://docs.oracle.com/javase/7/docs/jre/api/net/httpserver/spec/com/sun/net/httpserver/BasicAuthenticator.html), but I'm not sure how to add authorization (roles).

**Comment by alechenninger** [comment]
At least for my use case, I don't care about the roles so much. Does lightblue require them with auth?

**Comment by paterczm** [comment]
From what I'm seeing, BasicAuthenticator is extremely simple - just a user/pass to protect given resource. I'm not sure it creates a real principal that you can access in a lightblue hook. Requires further investigation.

If what HttpServer offers is not enough, then you need a real container. Arquillian (I have bad experience with it, but you got it to work, if I remember correctly?) or Docker. We have [integration tests example using Docker](https://github.com/lightblue-platform/lightblue-integration-tests), but a Docker lightblue image supporting basic auth would need to be prepared.

**Comment by paterczm** [comment]
It may create principal after all: http://docs.oracle.com/javase/7/docs/jre/api/net/httpserver/spec/com/sun/net/httpserver/Authenticator.Success.html

**Comment by alechenninger** [comment]
For now I think I may have found a workaround that isn't too bad, which lowers the priority on this for me. Thank you for looking into it!

**Comment by dcrissman** [comment]
LightblueRestTestHarness now uses Undertow which is a more robust http server than we were using before. I have not investigated this yet, but it allows for more complex configurations and might make this more of a possibility.

**Comment by dcrissman** [comment]
Actually, I have an example of using basic auth with Undertow [1]. Does that solve the need?

[1] https://github.com/lightblue-platform/lightblue-rest/blob/master/lightblue-rest-integration-test/src/test/java/com/redhat/lightblue/rest/logging/TestLoggingFilter.java#L110

**Comment by alechenninger** [comment]
That looks like there is a client request with basic auth. Do you have lightblue configured to require authentication in this test?

**Comment by dcrissman** [comment]
Yes - https://github.com/lightblue-platform/lightblue-rest/blob/master/lightblue-rest-integration-test/src/test/java/com/redhat/lightblue/rest/logging/TestLoggingFilter.java#L50

Here is where the work is actually done in the parent hierarcy: https://github.com/lightblue-platform/lightblue-rest/blob/master/lightblue-rest-integration-test/src/main/java/com/redhat/lightblue/rest/integration/LightblueRestTestHarness.java#L173

**Comment by alechenninger** [comment]
What is `LightblueRestTestClient`? Can I use this with tests that use the `LightblueExternalResource`?

**Comment by dcrissman** [comment]
> What is LightblueRestTestClient?

It is a little utility that I wrote to be able to test the rest layer without the lightblue-client library. I wouldn't expect it to be used much outside of lightblue-rest.

> Can I use this with tests that use the LightblueExternalResource?

No, that is part of the lightblue-client library.

**Comment by dcrissman** [comment]
My apologies, I think I misunderstood that last question. 

I don't think LightblueExternalResource exposes the functionality, but it exists on LightblueRestTestHarness, so it is possible.

**Comment by dcrissman** [comment]
@alechenninger Could you rebase https://github.com/lightblue-platform/lightblue-client/pull/268 ? If so, I would approve. I am working on exposing this functionality through LightblueExternalResource, but need your PR for testing. Thanks.

**Comment by dcrissman** [comment]
I worked some on this last night [1], it seems that auth will fail (correctly) when using invalid credentials but works just fine (incorrect) when passing no credentials. I wrote some unit tests to demonstrate this. I will need to continue to investigate as I have time. I am wondering if undertow needs to be somehow configured to not allow no auth.

Also did some work on the lightblue-client [2], but put it aside when I noticed the above issue.

[1] https://github.com/lightblue-platform/lightblue-rest/compare/master...dcrissman:authworking
[2] https://github.com/lightblue-platform/lightblue-client/compare/master...dcrissman:authworking
